explorer/interpreter/type_checker.cpp

```diff
@@ -2679,6 +2679,74 @@ auto TypeChecker::TypeCheckTypeExp(Nonnull<Expression*> type_expression,
   return type;
 }
 
+auto TypeChecker::TypeCheckTypeReturnDeclExp(
+    Nonnull<Expression*> type_expression, const ImplScope& impl_scope)
+    -> ErrorOr<Nonnull<const Value*>> {
+  CARBON_RETURN_IF_ERROR(TypeCheckExp(type_expression, impl_scope));
+  Nonnull<const Value*> type;
+  if (type_expression->kind() == ExpressionKind::CallExpression) {
```

Comment:
Note, this special-casing on CallExpression is what concerns me, I think a call to IsType should more be more precise.

This may not address a part of #1394 though if taking that approach; I'm not sure if there's still some issue with `match (true[true])`

Reply:
Just to explain, the error is not due to match(true[true]). Also the following code causes a crash
```carbon
fn test(){
   "a"["1"];
}
fn foo() -> test();
```
The reason is a bit more subtle:
The method **TypeChecker::TypeCheck(AST& ast) -> ErrorOr<Success> ** first executes DeclareDeclaration for the complete AST. When the declaration **fn foo()->test()** is visited, via DeclareCallable the method **TypeCheckTypeExp** is called for the ReturnType of the callable. Unfortunately this method calls InterpExp before the method body of fn test() has been checked and tries to execute it.
Therefore the new method **TypeCheckTypeReturnDeclExp** has been introduced to not execute the **InterpExp** call under certain conditions.

I hope it is understandable!

Reply:
After this change, what will happen with code like this?

```
fn test() -> i32 {
  "a"["1"];
}
fn foo() -> test();
```

Reply:
Sorry, I have thought too short! I think it is not possible to handle all X even with the current structure of the explorer. 

```carbon
fn test() -> X //
{
  "a"["1"];
}
fn foo() -> test();
```
The question is now what is the bug, is it the declaration **fn foo() -> test();** compiled or that "a"["1"] is not caught and crashed?

Reply:
I think for now we should disallow any return type expression that involves a function call, regardless of whether the function type-checks. We'll need to revisit that rule when we start adding metaprogramming support, but at that point we'll presumably have a metaprogramming design to guide us on how to handle cases like this.

Reply:
I think `-> test()` is the issue. Again, we haven't really discussed metaprogramming -- I'm going to say no metaprogramming for now.

I suspect we don't have any tests checking the behavior right now of whether `fn foo() -> test()` is working, but if we do it should be fine to switch them to disallow it.

Note, while this example may seem to work okay, I'm suggesting we shouldn't embark on it this way because metaprogramming quickly becomes complex. For a few examples:

Simple indirection:

```
fn test() -> Type {
  return ();
}

fn Main() -> test() {
  return 0;
}
```

Globals:

```
var x: Type = i32;

fn test() -> Type {
  // This should probably be invalid because of referencing a non-local variable.
  // But similar with `let` would probably be okay, right?
  return x;
}

fn Main() -> test() {
  return 0;
}
```

More indirection:

```
fn test1() -> Type {
  return i32;
}

fn test2() -> Type {
  return test1();
}

fn Main() -> test2() {
  return 0;
}
```

Random compilation (would this be allowed? it's non-deterministic):

```
fn test() -> Type {
  if (Rand()) {
    return ();
  } else {
    return i32;
  }
}

fn Main() -> test() {
  return 0;
}
```

Printing during compilation (would this be allowed? it's non-deterministic too, since printing is an external API call):

```
fn test() -> Type {
  Print("blah");
  return i32;
}

fn Main() -> test() {
  return 0;
}
```

Adding metaprogramming really requires more comprehensive design and testing, and I think we shouldn't enter into it here. Compile-time function calling is one metaprogramming approach, and while we may allow it in some manner, we may want something closer to the `constexpr` boundaries that C++ has. Providing something without that boundary doesn't seem worth the cost of sorting through the issues it creates.

Reply:
I haven't thought about the fix too much, but the answer is probably in constraining expressions that InterpExp would evaluate.

Reply:
No, it should not -- `F(i32)` is also compile-time metaprogramming. I think that test can be changed to a fail test, where `var v: F(i32)` is invalid.

Reply:
Thanks!